### PR TITLE
feat(push_rollout): `max_in_progress_updates` update api

### LINKED ISSUE #278: Add API to increase `max_in_progress_updates`
Right now we don't have any timeout for all the OTAOperation states besides `Pending`. This is because a device can, potentially, update its state only a long time after it actually changes it (e.g.: a device reboots and in that moment an internet fiber cable gets cut, resulting in it being able to update the state to `Success` only after some time).
This however means that all the slots for `max_in_progress_updates` could be occupied by devices waiting for status updates. To provide an escape hatch if the user wants the campaign to still make progress, we should allow to raise (not lower) the number of `max_in_progress_updates`, so new slots can become available.

### PR DESCRIPTION
**Pull request by noaccOS**
provide an api for updating push_rollout.
for now, only the field `max_in_progress_updates` is exposed through the graphql mutation.

includes edits to `push_rollout_executor` to handle the updates.

closes #278
re-implementation of #305
adapts and includes #325

### CODE REVIEW

File: backend/test/edgehog_web/schema/mutation/update_rollout_mechanism_test.exs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      )
+      |> Enum.into(%{})
+
+    variables = %{input: input}
+
+    conn = post(conn, api_path, query: @query, variables: variables)
+
+    json_response(conn, 200)
+  end
+end
```
Comment (szakhlypa) [inline backend/test/edgehog_web/schema/mutation/update_rollout_mechanism_test.exs:104]: Not sure about naming, but having `updatePushRollout` mutation I think better to rename file to `update_push_rollout_test.exs` and update tests to:
```suggestion
defmodule EdgehogWeb.Schema.Mutation.UpdatePushRolloutTest do
  use EdgehogWeb.ConnCase

  import Edgehog.UpdateCampaignsFixtures

  describe "updatePushRollout mutation" do
    setup do
      {:ok, update_campaign: update_campaign_fixture()}
    end

    test "succeeds to increase max_in_progress_updates", %{
      conn: conn,
      api_path: api_path,
      update_campaign: update_campaign
    } do
      max_in_progress_updates = update_campaign.rollout_mechanism.max_in_progress_updates + 1

      response =
        update_push_rollout_mutation(conn, api_path,
          update_campaign_id: update_campaign.id,
          max_in_progress_updates: max_in_progress_updates
        )

      update_campaign = response["data"]["updatePushRollout"]["updateCampaign"]
      assert update_campaign["rolloutMechanism"]["maxInProgressUpdates"] ==
               max_in_progress_updates
    end

    test "fails to reduce max_in_progress_updates", %{
      conn: conn,
      api_path: api_path,
      update_campaign: update_campaign
    } do
      response =
        update_push_rollout_mutation(conn, api_path,
          update_campaign_id: update_campaign.id,
          max_in_progress_updates: update_campaign.rollout_mechanism.max_in_progress_updates - 1
        )

      assert response["data"]["updateCampaign"] == nil
      assert [%{"status_code" => 422, "message" => message}] = response["errors"]
      assert message =~ "must be greater than or equal to its previous value"
    end

    test "fails when trying to use a non-existing update campaign", %{
      conn: conn,
      api_path: api_path
    } do
      response = update_push_rollout_mutation(conn, api_path, update_campaign_id: "123456")

      assert %{"errors" => [%{"status_code" => 404, "code" => "not_found"}]} = response
    end
  end

  @query """
  mutation UpdatePushRollout($input: UpdatePushRolloutInput!) {
    updatePushRollout(input: $input) {
      updateCampaign {
        rolloutMechanism {
          ... on PushRollout {
            maxInProgressUpdates
          }
        }
      }
    }
  }
  """
  defp update_push_rollout_mutation(conn, api_path, opts) do
    update_campaign_id =
      Absinthe.Relay.Node.to_global_id(
        :update_campaign,
        opts[:update_campaign_id],
        EdgehogWeb.Schema
      )

    input =
      Enum.into(opts, %{})
      |> Map.put(:update_campaign_id, update_campaign_id)

    variables = %{input: input}

    conn = post(conn, api_path, query: @query, variables: variables)

    json_response(conn, 200)
  end
end
```

**Review by szakhlypa — CHANGES_REQUESTED** [review]

File: backend/lib/edgehog/update_campaigns/push_rollout/executor.ex [hunk 2]
```diff
@@ -607,4 +616,19 @@ defmodule Edgehog.UpdateCampaigns.PushRollout.Executor do
...
+    new_data.available_slots > old_data.available_slots
+  end
+
+  defp update_rollout_mechanism(data, rollout) do
+    current_in_progress = data.in_progress_count
+    available_slots = Core.available_update_slots(rollout, current_in_progress)
+
+    %{
+      data
+      | rollout: rollout,
```
[the remark this answers is not in the document]
Comment (noaccOS) [reply backend/lib/edgehog/update_campaigns/push_rollout/executor.ex]: _if only this were a strongly typed language_

File: backend/lib/edgehog/update_campaigns.ex [hunk 3]
```diff
@@ -376,6 +377,32 @@ defmodule Edgehog.UpdateCampaigns do
...
+      |> Repo.update()
+
+    with {:ok, update_campaign} <- update_result do
+      update_campaign = preload_defaults_for_update_campaign(update_campaign)
+
+      _ = PubSub.publish!(:update_campaign_updated, update_campaign)
+
+      {:ok, update_campaign}
+    end
+  end
```
[the remark this answers is not in the document]
Comment (noaccOS) [reply backend/lib/edgehog/update_campaigns.ex]: done

File: backend/lib/edgehog/update_campaigns/push_rollout/executor.ex [hunk 4]
```diff
@@ -374,6 +373,16 @@ defmodule Edgehog.UpdateCampaigns.PushRollout.Executor do
   # events enqueued with the :next_event action. This means that we can be sure an :info event
   # or a timeout won't be handled, e.g., between a rollout and the handling of its error
 
+  def handle_event(:info, {:update_campaign_updated, update_campaign}, state, data) do
+    new_data = update_rollout_mechanism(data, update_campaign.rollout_mechanism)
+
+    if state == :wait_for_available_slots and new_slots_available?(data, new_data) do
```
[the remark this answers is not in the document]
Comment (noaccOS) [reply backend/lib/edgehog/update_campaigns/push_rollout/executor.ex]: makes sense, corrected

File: backend/lib/edgehog/pubsub.ex [hunk 5]
```diff
@@ -73,4 +81,6 @@ defmodule Edgehog.PubSub do
   defp topic_for_subject(%OTAOperation{id: id}), do: "ota_operations:#{id}"
   defp topic_for_subject({:ota_operation, id}), do: "ota_operations:#{id}"
   defp topic_for_subject(:ota_operations), do: "ota_operations:*"
+  defp topic_for_subject(%UpdateCampaign{id: id}), do: "update_campaign:#{id}"
+  defp topic_for_subject({:update_campaign, id}), do: "update_campaign:#{id}"
```
[the remark this answers is not in the document]
Comment (noaccOS) [reply backend/lib/edgehog/pubsub.ex]: yeah best to keep it consistent

File: backend/test/edgehog_web/schema/mutation/update_rollout_mechanism_test.exs (hunk 1 above)
Comment (noaccOS) [reply backend/test/edgehog_web/schema/mutation/update_rollout_mechanism_test.exs]: yup, forgot to check naming. thanks

File: backend/lib/edgehog_web/schema/update_campaigns_types.ex [hunk 6]
```diff
@@ -150,6 +150,20 @@ defmodule EdgehogWeb.Schema.UpdateCampaignsTypes do
...
+  """
+  input_object :push_rollout_update do
+    @desc """
+    The maximum number of in progress updates. The Update Campaign will have \
+    at most this number of OTA Operations that are started but not yet \
+    finished (either successfully or not).
+
+    Must be greater than (or equal to) the current value.
+    """
+    field :max_in_progress_updates, :integer
```
Comment (szakhlypa) [inline backend/lib/edgehog_web/schema/update_campaigns_types.ex:164]: Should we also make `:max_in_progress_updates` non nullable?
```suggestion
    field :max_in_progress_updates, non_null(:integer)
```

**Review by szakhlypa — CHANGES_REQUESTED** [review]

**Comment by rbino** [comment]
Closing this too, superseded by the Ash porting (the issue remains open, but this needs to be reimplemented on top of Ash)

**Pull request closed without merging**